### DISCUSSION TITLE
When will be able to specify `runs.using: node18` in action.yml?

### DISCUSSION THREAD

**Discussion by ezzatron** [actions/virtual-environments#7664] [General questions] [answered]
The default version of Node.js has apparently been 18.x [since 2023-02-13](https://github.com/actions/runner-images/issues/7002), Node.js 16.x will be end of life in 3 months on [2023-09-11](https://nodejs.dev/en/about/releases/), and [`@octokit/action` has recently been updated to require `"node": ">= 18"`](https://github.com/octokit/action.js/blob/6ed6998885a12908c1f9361080761f12492e6893/package.json#L91). But as of today, we still can't specify `node18` as the [runtime in `action.yml` metadata `runs.using` field](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-javascript-actions).

If you do specify something like this:

```yml
# action.yml
runs:
  using: node18
  main: dist/main.js
```

The action will fail with the following message:

```
Error: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''using: node18' is not supported, use 'docker', 'node12' or 'node16' instead.')
   at GitHub.Runner.Worker.ActionManifestManager.ConvertRuns(IExecutionContext executionContext, TemplateContext templateContext, TemplateToken inputsToken, String fileRelativePath, MappingToken outputs)
   at GitHub.Runner.Worker.ActionManifestManager.Load(IExecutionContext executionContext, String manifestFile)
```

This is a pretty poor experience.

I get that Node.js 16.x is not _technically_ end-of-life yet, and is still receiving security releases. But I think I speak on behalf of everyone who develops actions when I say that we **need** a decent window of time to test our actions with a new runtime. We don't want to be forced into doing our upgrades with such a small time window before end-of-life happens.

Could we please get an answer as to when we will be able to actually use Node.js 18.x? And can we have a discussion about reducing the lead time on new Node.js LTS versions in GitHub Actions, so that this situation doesn't repeat itself for Node.js 20.x?

See also #5429.

**Comment by vpolikarpov-akvelon** [comment, marked as answer]
Hi @ezzatron. Unfortunately, it's out of our control.
Node.js versions that may be used to execute custom actions are [enforced by runner itself](https://github.com/actions/runner/blob/main/src/Runner.Common/Util/NodeUtil.cs#L6). I believe it would be better to ask there.

I also found this discussion: https://github.com/actions/runner/issues/2619. I hope this helps.

**Reply by ezzatron** [reply]
Thanks for pointing me in the right direction, appreciate it 👍

**Reply by ezzatron** [reply]
For anyone else looking, there's also a discussion at https://github.com/orgs/community/discussions/53217 with the most recent news being a comment on 2023-05-20 by @fhammerl:

> With Node16 security updates coming to an end this October, we're of course planning on embedding a new Node.js version, but I can't give an exact date as of today.

It's also being speculated that we might get a bump straight from 16.x to 20.x.
